--- makefile.orig	2026-07-10 23:25:20.312240921 +0300
+++ makefile	2026-07-10 23:25:35.764187119 +0300
@@ -75,16 +75,23 @@
 # For C89, "-std=c89 -DLUA_USE_C89"
 # Note that Linux/Posix options are not compatible with C89
 # (For 32-bit, add option "-m32" to MYCFLAGS and MYLDFLAGS.)
-MYCFLAGS= $(LOCAL) -std=c99 -DLUA_USE_LINUX
-MYLDFLAGS= -Wl,-E
-MYLIBS= -ldl
-
-
-CC= gcc
-# (Optionally we can use -march=native -mno-avx512f.)
-CFLAGS= -Wall -O2 $(MYCFLAGS) -fno-stack-protector -fno-common
-AR= ar rc
-RANLIB= ranlib
+MYCFLAGS= $(LOCAL) -std=c99 \
+	-I../../../libc/musl/arch/x86_64 \
+	-I../../../libc/musl/arch/generic \
+	-I../../../libc/musl/obj/include \
+	-I../../../libc/musl/include
+MYLDFLAGS= -static
+MYLIBS=
+
+
+CC= clang --target=x86_64-linux-musl
+CFLAGS= -Wall -O2 -g -fno-pic $(MYCFLAGS) -fno-stack-protector -fno-common
+AR= llvm-ar rc
+RANLIB= llvm-ranlib
+LD= ld.lld
+LDFLAGS= -m elf_x86_64 -T ../../../init/linker.ld --oformat=elf64-x86_64 -nostdlib
+MUSL_CRT= ../../../libc/musl/lib/crt1.o
+MUSL_LIBC= ../../../libc/musl/lib/libc.a
 RM= rm -f
 
 
@@ -92,7 +99,7 @@
 # == END OF USER SETTINGS. NO NEED TO CHANGE ANYTHING BELOW THIS LINE =========
 
 
-LIBS = -lm
+LIBS =
 
 CORE_T=	liblua.a
 CORE_O=	lapi.o lcode.o lctype.o ldebug.o ldo.o ldump.o lfunc.o lgc.o llex.o \
@@ -121,8 +128,9 @@
 	$(AR) $@ $?
 	$(RANLIB) $@
 
-$(LUA_T): $(LUA_O) $(CORE_T)
-	$(CC) -o $@ $(MYLDFLAGS) $(LUA_O) $(CORE_T) $(LIBS) $(MYLIBS) $(DL)
+$(LUA_T): $(LUA_O) $(CORE_T) $(MUSL_CRT) $(MUSL_LIBC)
+	$(LD) $(LDFLAGS) $(MUSL_CRT) $(LUA_O) $(CORE_T) $(MUSL_LIBC) -o $@
+	llvm-size $@
 
 
 clean:
